### DISCUSSION TITLE
Dynamic routes with getStaticProps

### DISCUSSION THREAD

**Discussion by ipaymushkin** [vercel/next.js#15691] [Help]
Hello!
How to use `dynamic routes` with `getStaticProps`?
I have many user pages (username is unique). I am going to use username as html name (`<username> .html`).
When a new user is created I want to create a new route without rebooting.
I want to take full advantage of static generation.
When I use `getStaticProps `with dynamic routes, I get an error:
```
Error: getStaticPaths is required for dynamic SSG pages and is missing for '/[id]'.
Read more: https://err.sh/next.js/invalid-getstaticpaths-value
```
I can use `getStaticPaths`, but then I have to make a request for a list of users every time.

**Comment by alanqchen** [comment]
`getStaticPaths` is a requirement when using `getStaticProps` in a dynamic route. What looks to be useful to you is the [`fallback` option](https://nextjs.org/docs/basic-features/data-fetching#the-fallback-key-required) in `getStaticPaths`. Using this, you fetch and generate the initial list of user pages when you build, and when more users are added at runtime, the new pages will be generated when the page is requested.

I would also suggest taking a look at the [`revalidate` option](https://nextjs.org/docs/basic-features/data-fetching#the-fallback-key-required) in `getStaticProps`. This will have it so if someone requests `/<username>` before the user exists, it won't stay that way permanently. 

Using the above has some drawbacks: 

1. If the user page was not yet pre-generated, the loading time is slower to populate the page (only applies to the first request)
2. If the user page was already pre-generated before, but was not valid at the time, then it will take two requests for the updated page to appear (once to start the pre-generation in the background after the `revalidate` time has passed, and second to see the updated result)

If the above isn't acceptable then I might consider statically generating a "template" page and then requesting the user data on the client (i.e don't user `getStaticProps` or `getServerSideProps`).
